### ISSUE TITLE
Support webcal http auth & store credentials

### ISSUE THREAD

**Issue by tcitworld** [nextcloud/calendar#70]
Ideally : 
- Have a checkbox "Provide credentials" beneath the url input field
- Show login and password beneath when it's checked
- Save credentials in the app. *Secured enough ?*
- Perform the proxy request and provide these credentials in the headers

What do you think ?

**Comment by raimund-schluessler** [comment]
Regarding security/storing of credentials @LukasReschke should have look too I guess.

**Comment by georgehrke** [comment]
Well, you can already store credentials by putting them in the url with user:password@url,
The problem is, that a user with whom the calendar is shared is will see the credentials.
But as the webcal is fetched by the browser, you would not be able to share password protected webcal calendars

**Comment by tcitworld** [comment]
I wonder if webcal calendars should be shared anyway, since you can just give the link to other people.

**Comment by georgehrke** [comment]
Sharing it is still way easier than telling someone to copy and paste that url into the new subscription field.

**Comment by tcitworld** [comment]
> Well, you can already store credentials by putting them in the url with user:password@url

Right now ? Doesn't work for me, maybe because it seems http:// gets added on front.

**Comment by Tanarri** [comment]
I have also a problem, that I have to put ht eusername and password in the URL.
...and I have a non standard.port for my cloud.

This result in a URL für webdav:
webdav://username:password/example.to:1234/cloud/remote.php/dav/calendar.ics

Outlook can't get the calendar. Error:
"Der Internetkalender kann in Outlook nicht überprüft oder hinzugefügt werden. Setellen sie sicher, dass der Hyperlink ein gültiger Kalenderhyperlink ist.

Any suggestions? Or is this not supported (credentials + special port)

**Comment by georgehrke** [comment]
It's supposed to be user:pass@host:port/path

**Comment by treiher** [comment]
It doesn't work if I add https:// on front. Is only HTTP supported?

**Comment by wuethrich44** [comment]
I would like to subscribe to another Nextcloud calendar so I tried to subscribe with the following URL:

**Comment by tcitworld** [comment]
@wuethrich44 Not that I know of, but you should publish the calendar on the other end and then subscribe to it's public url.

**Comment by wuethrich44** [comment]
@tcitworld Ok that could be an approach but I don't have admin rights in the other server so I can't make the calendar public.

**Comment by tcitworld** [comment]
You don't need admin rights, just regular user access that you seem to have.

**Comment by treiher** [comment]
As stated in Issue #287 it should be possible to retrieve a WebCal calendar via HTTPS. For me it does not work, if I need authentication at the same time.

If I subscribe to
`https://USERNAME:PASSWORD@HOST/url/to/calendar`
the App changes the URL after creation to
`https://USERNAME@HOST/url/to/calendar`,
which obviously ends in a 
`The remote server did not give us access to the calendar (HTTP {403} error)`.

Am I doing something wrong? How can I subscribe to a calendar with authentication and HTTPS?

**Comment by wuethrich44** [comment]
I've observed the same behavior. I use the following link to share the calendar:

![image](https://cloud.githubusercontent.com/assets/6050639/21865814/eaa8961e-d847-11e6-8870-88cc8111ae73.png)

**Comment by elgaard** [comment]
For me , the password is intact as source in the oc_calendarsubscriptions table.
But not in the url used for the proxy.

**Comment by laeubi** [comment]
Maybe the calender can check for the 401 code and then ask the user for user/password?

**Comment by georgehrke** [comment]
> Maybe the calender can check for the 401 code and then ask the user for user/password?

Yes, that's what this ticket is about :) 

> Beside this it would be cool if subscribing has an option to sync the remote calendar into a local one so one is able to share this calendar with others by using a local one...

This would be an issue that belongs into the server repository.
See:
https://github.com/nextcloud/server/issues/1497
https://github.com/nextcloud/server/issues/1600

**Comment by guilhem-g** [comment]
> For me , the password is intact as source in the oc_calendarsubscriptions table.
But not in the url used for the proxy.

@elgaard  It's because the method `build `of the sabre\uri library contains : 
```php
$authority = $parts['user'] . '@' . $authority;
```
and should maybe be :
```php
$authority = $parts['user'] . ':' . $parts['pass'] . '@' . $authority;
```
I'v tried and i was able to import my calendar.

**Comment by tcitworld** [comment]
@pepea28 Would you care to open a PR to the https://github.com/sabre-io/uri/ repository ?

**Comment by rugk** [comment]
So what is the state of this issue? (i.e. that the password is removed when it is requested by the proxy) I see none of your PRs are merged, @tcitworld…

Or is there at least any workaround one can try?

**Comment by rugk** [comment]
FYI the Nextcloud "news" app does have a similar UI as proposed in the OP (with a field for username/password).

**Comment by protree** [comment]
I also need this feature. Could the according PRs please be merged?

**Comment by ChristophWurst** [comment]
> I also need this feature. Could the according PRs please be merged?

What PRs?

**Comment by tcitworld** [comment]
https://github.com/nextcloud/3rdparty/pull/143

**Comment by ChristophWurst** [comment]
It turns out that you can now actually use user:password with https links. I'm unable to reproduce the issue where sabre strips the password for https but not for http.

So the remaining part is storing the auth values separately but also the sharing. Even if we have a separate storage for credentials, you would currently have to expose them to the sharees so they are able to sync. Unless they have to submit their own password, but that kind of defeats the purpose of sharing.

**Comment by tcitworld** [comment]
That's odd, I don't see how this could work since not much has changed since then. Will try to reproduce.

> Even if we have a separate storage for credentials, you would currently have to expose them to the sharees so they are able to sync.

Subscriptions are exposed as regular calendars inside this app because they are cached on server, so we could make sure they don't expose auth information (currently probably accessible through `href` and default `displayname`).
We *could* do the same for any CalDAV client as long as the subscription URI contains credentials. In that case, we wouldn't need to worry about storing the auth values separately or how to handle sharing.

See https://github.com/nextcloud/server/blob/decb70b9ac6048e72305154af841271902f6b10c/apps/dav/lib/CalDAV/CalendarHome.php#L107-L116

**Comment by ChristophWurst** [comment]
Okay, so you would allow sharees to see the cached calendar, but they will be unable to refresh it, right?

**Comment by tcitworld** [comment]
No one can refresh the cached calendar subscriptions right now anyway.

This is either [defined by the calendar data itself](https://github.com/nextcloud/server/blob/2f26ff40e9aaabd1c8cc2167dea2158a22b1961f/apps/dav/lib/CalDAV/WebcalCaching/RefreshWebcalService.php#L309-L339) through `REFRESH-INTERVAL` or `X-PUBLISHED-TTL` icalendar keys, or fallback to [the value defined by the `calendarSubscriptionRefreshRate` app setting](https://github.com/nextcloud/server/blob/2f26ff40e9aaabd1c8cc2167dea2158a22b1961f/apps/dav/lib/BackgroundJob/RefreshWebcalJob.php#L90-L91) (defaults to one week).

**Comment by ChristophWurst** [comment]
Oh, then I simply read an outdated comment about the sync happening just in the browser. Thanks for the clarification.

**Comment by endo34** [comment]
Didn't see an answer while scrolling this thread: what if my username is an email address? The Webcal log-in should be: 

> user@example.com:password@http://example.com/webdav

I think the email ampersand (@) confuses NextCloud. I get an error that says: 

> Please enter a valid link (starting with http://, https://, webcal://, or webcals://)

**Comment by rugk** [comment]
If that is supposed to be a HTTP Auth (with username0email), the syntax is however indeed wrong.
Correct would be `http://user@example.com:password@example.com/webdav`. The protocol of course has to come first.

**Comment by endo34** [comment]
Of course- thank you. The problem remains that the @ is causing nextcloud to interpret the username as the service url. Probably a more clear explanation is:

>  http://username@username.com:password@service-address.com/webdav

Any suggestions?

**Comment by rightsaidfred99** [comment]
Does this method the username and password work with Synology based calendars? 

I can't seem to get it to work using the user+pword in the url

**Comment by rugk** [comment]
> Is it safe to expose a password like this??

Well… it's stored next to the URL in the DB or wherever it is shown (AFAIK when hovering over the calendar in the app you can also see the URL or so). As such, that needs to be taken into account. The (calendar) devs or so may also not be aware that there can be sensitive information embedded.
Other than that, it's not more or less safe than a separate input for username and password. Remember [HTTPS URLs are encrypted](https://stackoverflow.com/questions/499591/are-https-urls-encrypted). Having that URL in your browser history would potentially be worse in most threat models… :upside_down_face:

**Comment by mkrecek234** [comment]
@tcitworld HTTP Authentication does not seem to work reliably - I cannot add a calendar subscription in Nextcloud with such a link (which perfectly works in Google Chrome to download the ics files).

https://u.ser%40hadomain.de:smallBig%3B2@mail.hadomain.de/SOGo/dav/u.ser@hadomain.de/Calendar/personal.ics

I get Authentication failed in Nextcloud log, as it seems not to store the credentials properly.

**Comment by rokejulianlockhart** [comment]
For this, some kind of stock credential manager would be necessary, surely. I realize that this would be a lot to implement, but I'd love to be able to store and manage these credentials from Bitwarden.
